pyquil/quil.py

```diff
@@ -289,11 +289,24 @@ def inst(self, *instructions: InstructionDesignator) -> "Program":
 
             # Implementation note: these two base cases are the only ones which modify the program
             elif isinstance(instruction, DefGate):
-                defined_gate_names = [gate.name for gate in self._defined_gates]
-                if instruction.name in defined_gate_names:
-                    warnings.warn("Gate {} has already been defined in this program".format(instruction.name))
+                # If the gate definition differs from the current one, print a warning and replace it.
+                try:
+                    existing_defgate = next(gate for gate in self._defined_gates if gate.name == instruction.name)
+                    if (instruction.matrix.dtype == np.complex_) or (instruction.matrix.dtype == np.float_):
+                        if not np.allclose(existing_defgate.matrix, instruction.matrix):
+                            warnings.warn("Redefining gate {}".format(instruction.name))
+                            self._defined_gates = [
+                                gate if gate.name != instruction.name else instruction for gate in self._defined_gates
+                            ]
+                    else:
+                        if not np.all(existing_defgate.matrix == instruction.matrix):
+                            warnings.warn("Redefining gate {}".format(instruction.name))
+                            self._defined_gates = [
+                                gate if gate.name != instruction.name else instruction for gate in self._defined_gates
+                            ]
+                except Exception:
```

Comment:
What is the exception you expect to be raised here? It doesn't look like the `try` block raises any

Reply:
Oh, I guess it's raised by `next` on an empty iterator.

Reply:
I would drop the `try/except` and write it something like
```python
existing_defgate = next(..., None)  # default None
if existing_defgate is None:
    # append to defgates
elif (instruction.matrix.dtype == np.complex_) or (instruction.matrix.dtype == np.float_):
    # ...
elif not np.all(existing_defgate.matrix == instruction.matrix):
    # ...
```

Reply:
I didn't know `next()` had a second argument!